decred/tests/unit/dcr/test_dcrdata_unit.py

```diff
@@ -112,38 +109,54 @@ def test_checkoutput():
     # Amount is zero.
     tx = msgtx.TxOut()
     with pytest.raises(DecredError):
-        checkOutput(tx, 0)
+        dcrdata.checkOutput(tx, 0)
 
     # Amount is negative.
     tx = msgtx.TxOut(-1)
     with pytest.raises(DecredError):
-        checkOutput(tx, 0)
+        dcrdata.checkOutput(tx, 0)
 
     # Amount is too large.
     tx = msgtx.TxOut(txscript.MaxAmount + 1)
     with pytest.raises(DecredError):
-        checkOutput(tx, 0)
+        dcrdata.checkOutput(tx, 0)
 
     # Tx is dust output.
     script = ByteArray([opcode.OP_RETURN, opcode.OP_NOP])
     tx = msgtx.TxOut(value=1, pkScript=script)
     with pytest.raises(DecredError):
-        checkOutput(tx, 0)
+        dcrdata.checkOutput(tx, 0)
 
 
 class MockWebSocketClient:
-    def __init__(self, *a, **k):
+    def __init__(self, url, on_message, on_close, on_error):
+        self.on_message = on_message
+        self.on_close = on_close
+        self.on_error = on_error
         self.sent = []
+        self.received = []
 
     def send(self, msg):
-        try:
-            msg = json.loads(msg)
-        except json.JSONDecodeError:
-            pass
         self.sent.append(msg)
 
     def close(self):
-        pass
+        self.on_close(self)
+
+    def get_emitter(self):
+        def emitter(msg):
+            self.received.append(msg)
+
+        return emitter
+
+
+class TweakedDcrdataClient(DcrdataClient):
+    def __init__(self, url, emitter=None, monkeypatch=None):
+        if monkeypatch is None:
+            raise RuntimeError("Need monkeypatch")
+        monkeypatch.setattr(ws, "Client", MockWebSocketClient)
+        super().__init__(url, emitter)
+        self.psClient()
+        self.emitter = self.ps.get_emitter()
```

Comment:
This has to be done in this order to end up with incoming messages being recorded by the `MockWebSocketClient`.

Reply:
That is simpler, I like simpler. I did not do it that way because I mistakenly thought the `emitter` function could not be a method, but I tried your code and it works. Thanks.